[D2] ToastView Component UI 구현

kimkyunghun3 opened the pull request:
## 작업 내용
Component Toast View 구현
배경 색상, 폰트, 이미지는 임시로 넣은 것이니 참고부탁드립니다.


### 스크린샷
![Simulator Screen Shot - iPhone 12 - 2022-07-02 at 22 56 28](https://user-images.githubusercontent.com/52434820/177004096-8adf3906-4ab2-4036-9d34-ea339a8a6d14.png)

kimkyunghun3 committed:
:bulb: ToastViewStyle 구현 #5

kimkyunghun3 committed:
:bulb: ToastView 구현 #5

szzang-kr committed:
Merge branch 'develop' of github.com:mash-up-kr/GGiriGGiri-iOS into feature/kh/ComponentUI

# Conflicts:
#	Projects/DesignSystem/DesignSystem.xcodeproj/project.pbxproj

szzang-kr commented on Projects/DesignSystem/Sources/Component/ToastView/DDIPToastView.swift:
```diff
@@ -0,0 +1,87 @@
+        
+        self.addSubview(titleLabel)
+        self.addSubview(iconImageView)
+        self.addSubview(descriptionLabel)
```
❓ 이전 component를 만들때에도 반복되던 코드들인데, 어떻게 줄일수 있을까요?

kimkyunghun3 replied:
저는 보통 extension를 활용해서 addSubViews라는 함수를 만들고 이를 배열 형태로 여러개 받는 것을 하곤 했습니다.
```swift
extension UIView {
    func addSubViews(_ views: [UIView]) {
        for view in views {
            self.addSubview(view)
        }
    }
}

// 사용처
self.addSubViews([titleLabel, iconImageView, descriptionLabel])
```

이 방법을 의도하신건가요?
아니면 다른 것을 생각하신건가요? 저는 사실 이 방식을 선호하지 않습니다. 왜냐하면 이 `extension`는 은닉화되어있지 않아서 모든 곳에서 접근이 가능하다는 단점이 존재합니다. 하지만 사용처가 많다면 이는 은닉화와 편리성을 두고 트레이드오프를 생각해볼 것 같습니다.
그래서 보통 사용처가 적다면 `private extension`를 사용해서 사용하는 곳에서 은닉화하여 사용하곤 합니다! 혹시 어떤 것을 의도한 것인지 그리고 저의 의도가 적절한 방향인지 궁금합니다! ( 이부분은 아직 확실치 않아서 커밋은 해두지 않고 있겠습니다..!)

dvHuni replied:
> `extension`는 은닉화되어있지 않아서 모든 곳에서 접근이 가능하다는 단점이 존재합니다.

모든 `UIView`에서 해당 `extension`을 접근 할 수 있는 단점일까요? 해당 단점은 
```swift
extension UIView where BooProtocol { }
```
처럼 특정 프로토콜을 채택하는 경우만 접근 하도록 제한을 둘 수 있습니다.

szzang-kr replied:
처음 의도 했던 내용은 `extension`을 사용하여 중복된 코드를 줄이는 것이였습니다.

또한 앞서 말씀드린 것 처럼 은닉화는 extension을 사용하더라도 충분히 가능합니다.

kimkyunghun3 replied:
처음 말씀하신 것처럼 `UIView` `extension`에서 `protocol`를 채택하는 방식으로 생각해보았는데, `protocol`의 타입을 정하기가 애매하고 어떤 것을 넣어야할지 모르겠었습니다. 그래서 `protocol`를 만들고 `extension`에서 `where` 조건으로 `UIView`를 넣는 방식도 결국 `UIView`에서만 사용할 수 있으므로 은닉화를 시킨 것이 아닐까 싶어서 이렇게 구현해보았습니다. 적절하지 않다면 피드백 해주시면 감사하겠습니다!
```swift
protocol addViewsAble {}

extension addViewsAble where Self: UIView {
    func addSubViews(_ views: [UIView]) {
        for view in views {
            self.addSubview(view)
        }
    }
}
```

szzang-kr replied:
위에서 말씀드린 예시는 잘못되었네요 죄송합니다. 😓 

말씀해주신대로 해당 `protocol`, 그리고 `UIView` 타입인 경우에만 `addSubViews()`를 사용할 수 있습니다.

모든 `UIView`에 해당 메소드가 노출되지 않도록 하려는 의도가 맞을까요?

kimkyunghun3 replied:
맞습니다!! 그런 의도로 했습니다 그래서 모든 곳에 노출되지 않도록 은닉화를 했습니다!

szzang-kr commented on Projects/DesignSystem/Sources/Component/ToastView/DDIPToastView.swift:
```diff
@@ -0,0 +1,87 @@
+        setToastViewLayout()
+    }
+    
+    public func setDescriptionLabel() -> String {
+        let text = self.style.description.rawValue.components(separatedBy: "!")
+            return text[0] + "!" + "\n" + text[1]
+    }
```
🔴 천천히 생각해보고 반영 부탁드려요

1. 네이밍이 set~ 으로 시작하는데 `return`이 있어요. 다음 두가지 중 하나를 적용하면 좋을것 같아요
   2-1. set만 하고싶다면, `descriptionLabel` 자체에 텍스트를 셋팅하면 될 것 같아요.
      ``` swift
      public func setDescriptionLabel() -> String {
        var text = self.style.description.rawValue.components(separatedBy: "!")
        text = text[0] + "!" + "\n" + text[1]
        descriptionLabel.text = text
     }
      ```
    2-2 get~ 을 적용한다면 다음과 같이 적용할 수 있을것 같아요
      ``` swift
    public func getDescriptionLabel() -> String {
         let text = self.style.description.rawValue.components(separatedBy: "!")
         return text[0] + "!" + "\n" + text[1]
     }
      ```
2. 내부에 있는 style을 가지고 String을 return 하는데, public일 필요는 없을 것 같아요.

kimkyunghun3 replied:
https://tv.naver.com/v/4980432/list/267189
예전에 들었던 강의에서 스위프트 컨벤션에 getter가 없어서 get를 메서드에 사용하는 것을 지양하라고 배웠습니다. 이는 get 자체 의미가 애매하다고 들어서 이를 사용하지 않는 것이 좋다는 것을 배웠습니다!
그래서 get이 아닌 set를 이용해서 수정했습니다!
<img width="1033" alt="스크린샷 2022-07-03 오후 1 36 01" src="https://user-images.githubusercontent.com/52434820/177024818-b5e3f24e-a33d-4a27-bbe0-fe376932cab5.png">

```swift
  private func setDescriptionLabel() {
      let text = self.style.description.rawValue.components(separatedBy: "\n")
      var descriptionText: String
      if text.count == 2 {
          descriptionText = text[0] + "\n" + text[1]
          descriptionLabel.text = descriptionText
      } else {
          descriptionText = text[0]
          descriptionLabel.text = descriptionText
      }
  }
```

szzang-kr commented on Projects/DesignSystem/Sources/Component/ToastView/DDIPToastView.swift:
```diff
@@ -0,0 +1,87 @@
+    
+    public func setDescriptionLabel() -> String {
+        let text = self.style.description.rawValue.components(separatedBy: "!")
+            return text[0] + "!" + "\n" + text[1]
```
❓ 함수의 내용이 "!" 뒤에 "\n"을 붙이는것인데, 텍스트 자체에 "\n"을 붙이면 안되는 이유가 있을까요?

kimkyunghun3 replied:
아하 가능할거같습니다! 수정해보았는데,  "\n" 로 분기처리 해도 될거 같습니다!
그리고 한 줄 짜리 description도 있어서 분기로 처리해서 에러가 나지 않도록 수정했습니다. 
```swift
  public func setDescriptionLabel() -> String {
      let text = self.style.description.rawValue.components(separatedBy: "\n")
      if text.count == 2 {
          return text[0] + "\n" + text[1]
      } else {
          return text[0]
      }
  }
```

szzang-kr replied:
처음 생각했던 내용은 description 텍스트 자체에 "\n"을 삽입해도 괜찮지 않을까? 였습니다. 😓 설명을 잘 못해드린것 같네요.

```swift
case save = "기프티콘이 갤러리에 저장되었어요."
case register = "뿌리기 등록이 완료되었어요!\n설정한 마감 시간이 지나면 자동적으로 기프티콘이 전달됩니다."
case apply = "응모가 완료되었어요!\n응모 결과는 마감 후에 확인할 수 있어요."
```

텍스트를 위처럼 사용하면 필요 없는 함수가 될 것 같아요

kimkyunghun3 replied:
아하 그러네요! 불필요한 메서드는 지우도록 하겠습니다!

szzang-kr commented on Projects/DesignSystem/Sources/Component/ToastView/DDIPToastView.swift:
```diff
@@ -0,0 +1,87 @@
+//
+
+import UIKit
+import PinLayout
```
🔴 사용하지 않는 라이브러리 혹은 프레임워크는 지워주세요

kimkyunghun3 replied:
앗 바로 지웠습니다!

szzang-kr commented on Projects/DesignSystem/Sources/Component/ToastView/DDIPToastViewStyle.swift:
```diff
@@ -0,0 +1,37 @@
+import UIKit
+
+public struct DDIPToastViewStyle {
+    public enum Title: String {
+        case save = "저장 완료"
+        case register = "등록 완료"
+        case apply = "응모 완료"
+    }
+    
+    public enum Description: String {
+        case save = "기프티콘이 갤러리에 저장되었어요."
+        case register = "뿌리기 등록이 완료되었어요! 설정한 마감 시간이 지나면 자동적으로 기프티콘이 전달됩니다."
+        case apply = "응모가 완료되었어요! 응모 결과는 마감 후에 확인할 수 있어요."
+    }
```
❓ Title과 Description이 같은 case를 가지고 있어요. 하나로 사용할 수 있는 방법이 있을 것 같은데 어떤것이 있을까요?

kimkyunghun3 replied:
```swift
  public enum ToastViewOptions {
      case save, register, apply
      
      var info: (title: String, description: String) {
          switch self {
          case .save:
              return ("저장완료", "기프티콘이 갤러리에 저장되었어요.")
          case .register:
              return ("등록 완료", "뿌리기 등록이 완료되었어요!\n 설정한 마감 시간이 지나면 자동적으로 기프티콘이 전달됩니다.")
          case .apply:
              return ("응모 완료", "응모가 완료되었어요!\n 응모 결과는 마감 후에 확인할 수 있어요.")
          }
      }
  }
```

연산 프로퍼티와 튜플을 활용해서 하나로 사용할 수 있도록 했습니다! 네이밍이 근데 조금 애매한거같기두하네요 
사용처에서 title이 두번반복되지 않도록 titleOption 요렇게 변경해서 했는데 괜찮을까요?
```swift
self.titleLabel.text = self.style.titleOption.info.title
```

szzang-kr reviewed (changes requested):
수고하셨어용 👍

AhnSangHee commented on Projects/DesignSystem/Sources/Component/ToastView/DDIPToastViewStyle.swift:
```diff
@@ -0,0 +1,37 @@
+        case save = "기프티콘이 갤러리에 저장되었어요."
+        case register = "뿌리기 등록이 완료되었어요! 설정한 마감 시간이 지나면 자동적으로 기프티콘이 전달됩니다."
+        case apply = "응모가 완료되었어요! 응모 결과는 마감 후에 확인할 수 있어요."
+    }
```
🔴 Title, Description이 동일한 case를 갖고 있는데, 중복을 제거하는 것이 좋을 것 같아요!

kimkyunghun3 replied:
```swift
  public enum ToastViewOptions {
      case save, register, apply
      
      var info: (title: String, description: String) {
          switch self {
          case .save:
              return ("저장완료", "기프티콘이 갤러리에 저장되었어요.")
          case .register:
              return ("등록 완료", "뿌리기 등록이 완료되었어요!\n 설정한 마감 시간이 지나면 자동적으로 기프티콘이 전달됩니다.")
          case .apply:
              return ("응모 완료", "응모가 완료되었어요!\n 응모 결과는 마감 후에 확인할 수 있어요.")
          }
      }
  }
```

연산 프로퍼티와 튜플을 활용해서 하나로 사용할 수 있도록 했습니다! 네이밍이 근데 조금 애매한거같기두하네요 
사용처에서 title이 두번반복되지 않도록 titleOption 요렇게 변경해서 했는데 괜찮을까요?
```swift
self.titleLabel.text = self.style.titleOption.info.title
```

AhnSangHee commented on Projects/DesignSystem/Sources/Component/ToastView/DDIPToastViewStyle.swift:
```diff
@@ -0,0 +1,37 @@
+    
+    public let title: Title
+    public let description: Description
+    public let imageIcon: String
```
❓imageIcon은 이미지 아이콘의 이름을 나타내는 것인가요?

kimkyunghun3 replied:
맞습니다! 사용처를 보면 현재에는 customImage가 없어서` UIImage(systemName: self.style.imageIcon)` 처럼 되어있지만, 나중에는 가져올 이미지의 아이콘 이름을 사용할거라 이렇게 네이밍을 지었습니다!

AhnSangHee reviewed (changes requested):
수고하셨습니다🙂

kimkyunghun3 committed:
:recycle: description 분기 처리 방식 변경

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

kimkyunghun3 committed:
:recycle: set 메서드 네이밍에 따른 return 없도록 내부 수정

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

kimkyunghun3 committed:
Merge remote-tracking branch 'origin/feature/kh/ComponentUI' into feature/kh/ComponentUI

kimkyunghun3 committed:
:recycle: 동일한 case 다른 enum 하나로 통합하도록 변경 #5

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

szzang-kr commented on Projects/DesignSystem/Sources/Component/ToastView/DDIPToastView.swift:
```diff
@@ -46,7 +46,7 @@ public class DDIPToastView: UIView {
     
     private func setUI() {
-        self.titleLabel.text = self.style.title.rawValue
+        self.titleLabel.text = self.style.titleOption.info.title
```
style에서 title을 return 하도록 property를 만들면 titleOption에 접근하지 않아도 될 것 같아요.

이렇게 만들면 ToastViewOptions을 은닉할 수 있겠죠?

```swift
self.titleLabel.text = self.style.title
```

kimkyunghun3 replied:
```swift
  private let titleOption: ToastViewOptions
  private let descriptionOption: ToastViewOptions
  private let imageIcon: String
  
  public var title: String {
      return titleOption.info.title
  }
  
  public var description: String {
      return titleOption.info.description
  }
  
  public var image: String {
      return self.imageIcon
  }
```

이렇게 해서 ToastViewOptions에 접근하지 않고 프로퍼티를 통해 은닉화했습니다! 이 방식을 말씀하신건가요?!

szzang-kr commented on Projects/DesignSystem/Sources/Component/ToastView/DDIPToastViewStyle.swift:
```diff
@@ -9,29 +9,32 @@
-    
-    public let title: Title
-    public let description: Description
+    public let titleOption: ToastViewOptions
+    public let descriptionOption: ToastViewOptions
```
titleOption과 description을 public으로 만들어서 접근 할 수 있게 한 이유가 있을까요?

요런 부분에서 은닉화가 필요해요. ToastViewOptions 전체를 알지 않아도 Toast에 사용될 title, description, image만 알면 되니까요


사용하는 예시를 통해 어떻게 구현할 수 있을 지 참고해 보세요
```swift
titleLabel.text = style.title
descriptionLabel.text = style.description
imageIcon.image = style.imageIcon
```

kimkyunghun3 replied:
위에 댓글로 말씀드렸습니다! 체이닝처럼 계속 들어가는 것은 좋지 않는 것 같습니다! 그래서 말씀해주신 방향으로 수정해보았습니다!

kimkyunghun3 committed:
:recycle: 불필요한 로직 제거 #5

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

kimkyunghun3 committed:
:recycle: ToastView title, description, image 접근 은닉화 #5

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

kimkyunghun3 committed:
:recycle: addSubView를 배열로 처리하도록 extension 구현 및 은닉화

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

szzang-kr commented on Projects/DesignSystem/Sources/Component/Extension/UIView+.swift:
```diff
@@ -0,0 +1,19 @@
+
+import UIKit
+
+protocol addViewsAble {}
```
`protocol` 이름은 대문자로 시작해주세요, 또한 ~able은 소문자로 지정합니다.
ex. Decodable, Encodable

```swift
protocol AddSubViewable { 
    func addSubViews(_ views: [UIView])
}

extension AddSubViewable where Self: UIView {
    func addSubViews(_ views: [UIView]) {
         for view in views {
             self.addSubview(view)
         }
     }
}

public class DDIPToastView: UIView, AddSubViewable { }

```

kimkyunghun3 replied:
```swift
protocol AddViewsable {}

extension AddViewsable where Self: UIView {
    func addSubViews(_ views: [UIView]) {
        for view in views {
            self.addSubview(view)
        }
    }
}
```
프토로콜 네이밍 컨벤션에 맞게 수정했습니다! 감사합니다!

kimkyunghun3 committed:
:recycle: 프로토콜 네이밍 변경 #5

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################
